qbraid_qir/qasm3/visitor.py

```diff
@@ -319,19 +589,102 @@ def _visit_basic_gate_operation(self, operation: qasm3_ast.QuantumGate) -> None:
+
+        # Use existing gate mapping logic but with profile awareness
+        gate_map = {
+            "h": qis.h if self._profile.restrictions.prefer_qis_over_native else qis.h,
+            "x": qis.x,
+            "y": qis.y,
+            "z": qis.z,
+            "s": qis.s,
+            "sdg": qis.s_adj,
+            "t": qis.t,
+            "tdg": qis.t_adj,
+            "cx": qis.cx,
+            "cnot": qis.cx,
+            "cz": qis.cz,
+            "ccx": qis.ccx,
+            "swap": qis.swap,
+            "rx": qis.rx,
+            "ry": qis.ry,
+            "rz": qis.rz,
+        }
 
-        for i in range(0, len(op_qubits), op_qubit_count):
-            # we apply the gate on the qubit subset linearly
-            qubit_subset = op_qubits[i : i + op_qubit_count]
-            if op_parameters is not None:
-                qir_func(self._builder, *op_parameters, *qubit_subset)
+        if op_name in gate_map:
+            qir_func = gate_map[op_name]
+            if op_name in ["rx", "ry", "rz"]:
+                op_parameters = self._get_op_parameters(operation)
+                if op_parameters:
+                    qir_func(self._builder, *op_parameters, *op_qubits)
+                else:
+                    raise_qasm3_error(f"Parametric gate {op_name} requires parameters")
             else:
-                qir_func(self._builder, *qubit_subset)
+                qir_func(self._builder, *op_qubits)
+        elif op_name == "id":
+            # Identity gate implementation
+            qubit = op_qubits[0]
+            qis.x(self._builder, qubit)
+            qis.x(self._builder, qubit)
+        else:
+            # Use the mapping system
+            try:
+                qir_func, expected_qubit_count = map_qasm_op_to_pyqir_callable(op_name)
+                if len(op_qubits) != expected_qubit_count:
+                    raise_qasm3_error(
+                        f"Gate {op_name} expects {expected_qubit_count} qubits, got {len(op_qubits)}"
+                    )
+
+                op_parameters = self._get_op_parameters(operation)
+                is_parametric = op_name in PYQIR_ONE_QUBIT_ROTATION_MAP or op_name in [
+                    "xx",
+                    "xy",
+                    "yy",
+                    "zz",
+                    "pswap",
+                    "cp",
+                    "cphaseshift",
+                    "cp00",
+                    "cphaseshift00",
+                    "cp01",
+                    "cphaseshift01",
+                    "cp10",
+                    "cphaseshift10",
+                    "ms",
+                    "prx",
+                ]
+
+                if is_parametric:
+                    if not op_parameters:
+                        raise_qasm3_error(f"Parametric gate {op_name} requires parameters")
+
+                    # Handle special parameter cases
+                    if op_name == "ms" and len(op_parameters) != 3:
```

Comment:
You do not need to account for such validations here as they are [already raised inside pyqasm](https://github.com/qBraid/pyqasm/blob/main/src/pyqasm/visitor.py#L826) (and many more). 
Essentially, during the conversion from qasm to qir we can be sure that the program is semantically valid and proceed ahead.